test/stage1/behavior/bugs/7140.zig

```diff
@@ -0,0 +1,20 @@
+const P = struct {
+    fn print() void {}
+};
+
+const print = P.print;
```

Comment:
Both `P.print` and `S.print` are redefinitions of this and should give compile errors.

Reply:
That was my first thought, but great chunks of the stdlib failed to compile meaning that people rely on some kind of (unwritten) scoping rules.

Since documentation is for weak (and other languages) I went for this minimal solution that makes more sense than the status quo.

Reply:
I think the documentation is quite clear about this:
> Shadowing
> It is never allowed for an identifier to "hide" another one by using the same name:

We definitely shouldn't be adding weird scoping rules just to avoid breaking the stdlib pre 1.0.

Reply:
> We definitely shouldn't be adding weird scoping rules just to avoid breaking the stdlib pre 1.0.

You're missing the forest for the trees, this PR fixes the weird unwritten scoping rule by enforcing the lookup order.

> It is never allowed for an identifier to "hide" another one by using the same name:

If that's meant to be applied on declarations _and_ variables I'll be happy to add the missing check.

Reply:
>You're missing the forest for the trees, this PR fixes the weird unwritten scoping rule by enforcing the lookup order.

It's more of a bug in the redeclaration check than an unwritten rule.

> If that's meant to be applied on declarations and variables I'll be happy to add the missing check.

That's how it is demonstrated in the [docs](https://ziglang.org/documentation/master/#toc-Shadowing). It doesn't even use the word declaration to avoid any confusion between declarations and variable _declarations_. Only thing that could be clarified is that container fields are not declarations.

Reply:
> It doesn't even use the word declaration to avoid any confusion between declarations and variable declarations.

You're confusing the lack of details for nuance, the examples you're quoting are _variable declarations_ and there's not a single word on what happens for declarations, fields and other kind of declarations one may run into.